File: src/main/java/com/bdev/hogwarts_api/service/teacher/TeacherServiceImpl.java
package com.bdev.hogwarts_api.service.teacher;

import com.bdev.hogwarts_api.dao.TeacherDao;
import com.bdev.hogwarts_api.data.converter.teacher.TeacherDtoConverter;
import com.bdev.hogwarts_api.data.converter.teacher.TeacherModelConverter;
import com.bdev.hogwarts_api.data.dto.teacher.Teacher;
import com.bdev.hogwarts_api.utils.EncodingUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

import static java.lang.String.format;
import static java.util.stream.Collectors.toList;

@Service
public class TeacherServiceImpl implements TeacherService {
    @Autowired
    private TeacherDao teacherDao;

    @Override
    public boolean exists(long id) {
        return teacherDao.exists(id);
    }

    @Override
    public List<Teacher> getAllTeachers() {
        return teacherDao.findAll()
                .stream()
                .map(TeacherModelConverter::convert)
                .collect(toList());
    }

    @Override
    public Optional<Teacher> getTeacherById(long id) {
        return Optional.ofNullable(
                teacherDao.findOne(id)
        ).map(TeacherModelConverter::convert);
    }

    @Override
    public Optional<Teacher> getTeacherByLogin(String login) {
        return Optional.ofNullable(
                teacherDao.findFirstByLogin(EncodingUtils.toBase64(login))
        ).map(TeacherModelConverter::convert);
    }

    @Override
    public long createTeacher(Teacher teacher) {
        if (teacher.getId() != null) {
            throw new RuntimeException("Teacher id should be null during creation");
        }

        return teacherDao.save(TeacherDtoConverter.convert(teacher)).getId();
    }

    @Override
    public void editTeacher(Teacher teacher) {
        if (!teacherDao.exists(teacher.getId())) {
            throw new RuntimeException(format("Teacher with id '%d' does not exist", teacher.getId()));
        }

        teacherDao.save(TeacherDtoConverter.convert(teacher));
    }

    @Override
    public void deleteTeacher(long id) {
        if (!teacherDao.exists(id)) {
            throw new RuntimeException(format("Teacher with id '%d' does not exist", id));
        }

        teacherDao.delete(id);
    }
}

File: src/main/java/com/bdev/hogwarts_api/service/teacher/TeacherServiceImpl.kt
package com.bdev.hogwarts_api.service.teacher

import com.bdev.hogwarts_api.dao.TeacherDao
import com.bdev.hogwarts_api.data.converter.teacher.TeacherDtoConverter
import com.bdev.hogwarts_api.data.converter.teacher.TeacherModelConverter
import com.bdev.hogwarts_api.data.dto.teacher.Teacher
import com.bdev.hogwarts_api.utils.EncodingUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.lang.String.format

@Service
class TeacherServiceImpl : TeacherService {
    @Autowired
    private lateinit var teacherDao: TeacherDao

    override fun getAllTeachers(): List<Teacher> {
        return teacherDao
                .findAll()
                .map { TeacherModelConverter.convert(it) }
    }

    override fun exists(id: Long): Boolean {
        return teacherDao.exists(id)
    }

    override fun getTeacherById(id: Long): Teacher? {
        return teacherDao
                .findOne(id)
                ?.let{ TeacherModelConverter.convert(it) }
    }

    override fun getTeacherByLogin(login: String): Teacher? {
        return teacherDao
                .findFirstByLogin(EncodingUtils.toBase64(login))
                ?.let{ TeacherModelConverter.convert(it) }
    }

    override fun createTeacher(teacher: Teacher): Long {
        if (teacher.id != null) {
            throw RuntimeException("Teacher id should be null during creation")
        }

        return teacherDao.save(TeacherDtoConverter.convert(teacher)).id ?: throw RuntimeException()
    }

    override fun editTeacher(teacher: Teacher) {
        if (!teacherDao.exists(teacher.id)) {
            throw RuntimeException(format("Teacher with id '%d' does not exist", teacher.id))
        }

        teacherDao.save(TeacherDtoConverter.convert(teacher))
    }

    override fun deleteTeacher(id: Long) {
        if (!teacherDao.exists(id)) {
            throw RuntimeException(format("Teacher with id '%d' does not exist", id))
        }

        teacherDao.delete(id)
    }
}
